core/common/src/internal/format/parser/ParserOperation.kt

```diff
@@ -153,101 +154,149 @@ internal class TimeZoneParserOperation<Output>(
                     }
 
-                    State.AFTER_HOUR -> when {
-                        input[index] == ':' -> {
-                            index++
-                            if (validateTimeComponent(2)) State.AFTER_COLON_MINUTE else State.INVALID
-                        }
+                    else -> State.INVALID
+                }
 
-                        validateTimeComponent(2) -> State.AFTER_MINUTE
-                        else -> State.INVALID
-                    }
+                State.AFTER_SIGN -> when {
+                    validateTimeComponent(2) -> State.AFTER_HOUR
+                    validateTimeComponent(1) -> State.END
+                    else -> State.INVALID
+                }
 
-                    State.AFTER_MINUTE -> when {
-                        validateTimeComponent(2) -> State.END
-                        else -> State.INVALID
+                State.AFTER_HOUR -> when {
+                    input[index] == ':' -> {
+                        index++
+                        if (validateTimeComponent(2)) State.AFTER_COLON_MINUTE else State.INVALID
                     }
 
-                    State.AFTER_COLON_MINUTE -> when {
-                        input[index] == ':' -> {
-                            index++
-                            if (validateTimeComponent(2)) State.END else State.INVALID
-                        }
+                    validateTimeComponent(2) -> State.AFTER_MINUTE
+                    else -> State.INVALID
+                }
+
+                State.AFTER_MINUTE -> when {
+                    validateTimeComponent(2) -> State.END
+                    else -> State.INVALID
+                }
 
-                        else -> State.INVALID
+                State.AFTER_COLON_MINUTE -> when {
+                    input[index] == ':' -> {
+                        index++
+                        if (validateTimeComponent(2)) State.END else State.INVALID
                     }
 
-                    State.END, State.INVALID -> break
+                    else -> State.INVALID
                 }
+
+                State.END, State.INVALID -> break
             }
+        }
+
+        return if (state == State.END) index else lastValidIndex
+    }
+}
+
+internal class NamedTimeZoneParserOperation<Output>(
+    setter: AssignableField<Output, String>
+) : TimeZoneParserOperation<Output>(setter) {
+    private enum class State {
+        START,
+        IN_PART,
+        AFTER_SLASH
+    }
+
+    override fun validateTimeZone(input: CharSequence, startIndex: Int): Int {
+        fun Char.isTimeZoneInitial() = isAsciiLetter() || this == '.' || this == '_'
+        fun Char.isTimeZoneChar() = isTimeZoneInitial() || isAsciiDigit() || this == '-' || this == '+'
+
+        var index = startIndex
+        var state = State.START
+
+        while (index < input.length) {
+            val currentChar = input[index]
+            state = when (state) {
+                State.START -> when {
```

Comment:
`START` and `AFTER_SLASH` states look like they can be merged into one, and then the state machine becomes a `var inPart : Boolean = false`.